File: frontend/src/auth/ids.ts
// ID / code generation (plan §5.i, §10). Pure, dependency-free, unit-testable.
import type { StudentCode, TeacherId, TeacherKind } from "./types";

export const ANIMALS = [
  "KESTREL", "FALCON", "OTTER", "LYNX", "MOTH", "HERON", "VIPER", "CRANE",
  "WREN", "ORYX", "IBIS", "MERLIN", "FINCH", "ROOK", "JAY", "KITE", "LARK",
];

// Ambiguity-free alphabet (no O/0, I/1) so codes are easy to read aloud.
const ALPHA = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";

function randInt(n: number): number {
  return Math.floor(Math.random() * n);
}

function token(len: number): string {
  let s = "";
  for (let i = 0; i < len; i++) s += ALPHA[randInt(ALPHA.length)];
  return s;
}

export function makePin(): string {
  return String(1000 + randInt(9000));
}

const SUBJECT_PREFIX: Record<string, string> = { Maths: "MATH", Science: "SCI" };

/** Admin issues a teacher credential scoped to a role. */
export function issueTeacherId(params: {
  kind: TeacherKind;
  subject: string;
  sections: string[]; // class teacher passes exactly one
}): TeacherId {
  const { kind, subject, sections } = params;
  if (sections.length === 0) throw new Error("at least one section required");
  if (kind === "class" && sections.length !== 1) {
    throw new Error("a class teacher owns exactly one section");
  }
  const prefix = SUBJECT_PREFIX[subject] ?? "SUBJ";
  const secPart = sections.map((s) => s.replace(/^7/, "")).join("");
  const id = `${prefix}-${secPart}-${token(4)}`;
  const label =
    kind === "class"
      ? `Class ${sections[0]} · all students`
      : `${subject} · ${sections.join(" ")}`;
  return {
    id,
    pin: makePin(),
    kind,
    subject,
    sections,
    label,
    issuedAt: new Date().toISOString(),
  };
}

const TEACHER_ID_RE = /^[A-Z]+-[A-Z0-9]+-[A-Z0-9]{4}$/;

export function isTeacherIdShape(id: string): boolean {
  return TEACHER_ID_RE.test(id.trim().toUpperCase());
}

/** Student self-creates a code from their chosen call-sign + section. */
export function makeStudentCode(params: {
  classLevel: string;
  section: string;
  subject: string;
  avatarId: number;
}): StudentCode {
  const { classLevel, section, subject, avatarId } = params;
  const animal = ANIMALS[avatarId % ANIMALS.length];
  const code = `${animal}·${1 + randInt(9)}${section}`;
  return {
    code,
    pinSet: false,
    classLevel,
    section,
    subject,
    avatarId,
    createdAt: new Date().toISOString(),
  };
}

export function isPinShape(pin: string): boolean {
  return /^\d{4}$/.test(pin);
}

File: backend/app/ids.py
"""ID / code generation (mirror of frontend/src/auth/ids.ts, plan §5.i)."""

from __future__ import annotations

import secrets

ANIMALS = [
    "KESTREL", "FALCON", "OTTER", "LYNX", "MOTH", "HERON", "VIPER", "CRANE",
    "WREN", "ORYX", "IBIS", "MERLIN", "FINCH", "ROOK", "JAY", "KITE", "LARK",
]
_ALPHA = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"  # no ambiguous O/0, I/1
_SUBJECT_PREFIX = {"Maths": "MATH", "Science": "SCI"}


def _token(n: int) -> str:
    return "".join(secrets.choice(_ALPHA) for _ in range(n))


def make_pin() -> str:
    return f"{secrets.randbelow(9000) + 1000}"


def make_student_code(section: str, avatar_id: int = 0) -> str:
    """Mint an anonymous call-sign.

    The animal is drawn at random and is *not* a function of ``avatar_id``.
    It used to be ``ANIMALS[avatar_id % len(ANIMALS)]``, which looked harmless
    but meant the call-sign was decided by the avatar picker — and that picker
    defaults to 0. In practice almost nobody changes it, so almost every child
    in a sitting came out a KESTREL: sixteen of the seventeen names were never
    issued, and a class full of near-identical codes is exactly how a child
    ends up typing someone else's.

    ``avatar_id`` is kept in the signature because callers pass it positionally
    and it stays meaningful as the child's chosen tile colour; it simply has no
    say in their name.
    """
    animal = secrets.choice(ANIMALS)
    # a 2-digit tail keeps codes memorable while making collisions negligible
    return f"{animal}·{secrets.randbelow(9) + 1}{section}{secrets.randbelow(90) + 10}"


def issue_teacher_id(kind: str, subject: str, sections: list[str]) -> str:
    if not sections:
        raise ValueError("at least one section required")
    if kind == "class" and len(sections) != 1:
        raise ValueError("a class teacher owns exactly one section")
    prefix = _SUBJECT_PREFIX.get(subject, "SUBJ")
    sec_part = "".join(s.lstrip("7") for s in sections)
    return f"{prefix}-{sec_part}-{_token(4)}"


def teacher_label(kind: str, subject: str, sections: list[str]) -> str:
    if kind == "class":
        return f"Class {sections[0]} · all students"
    return f"{subject} · {' '.join(sections)}"
